File: bst_avl/bst_avl_skeleton/lib/bst.rb
require 'byebug'


class BSTNode
  attr_accessor :left, :right
  attr_reader :value

  def initialize(value)
    @left = nil
    @right = nil
    @value = value
  end
end

class BinarySearchTree
  def initialize
    @root = nil
  end

  def insert(value)
    @root.nil? ? @root = BSTNode.new(value) :
    self.class.insert!(@root, value)
  end

  def find(value)
    BinarySearchTree.find!(@root, value)
  end

  def inorder

  end

  def postorder

  end

  def preorder

  end

  def height
    BinarySearchTree.height!(@root)
  end

  def min
  end

  def max
  end

  def delete(value)
    @root = BinarySearchTree.delete!(@root, value)
  end

  def self.insert!(node, value)
    return BSTNode.new(value) if node.nil?

    if value <= node.value
      node.left = BinarySearchTree.insert!(node.left, value)
    else
      node.right = BinarySearchTree.insert!(node.right, value)
    end

    node

  end

  def self.find!(node, value)
    return nil if node.nil?
    return node if node.value == value

    value < node.value ?
    BinarySearchTree.find!(node.left, value) :
    BinarySearchTree.find!(node.right, value)

  end

  def self.preorder!(node)

  end

  def self.inorder!(node)
    return [] unless node

    arr = []
    arr += BinarySearchTree.inorder!(node.left) if node.left
    arr << node.value
    arr += BinarySearchTree.inorder!(node.right) if node.right

    arr

  end

  def self.postorder!(node)

  end

  def self.height!(node)
    return -1 if node.nil?
    return 0 if node.left.nil? && node.right.nil?

    min_height = 0
    max_height = 0

    loop_node = node

    unless node.left.nil?
      loop do
        break if loop_node.left.nil?
        min_height += 1
        loop_node = loop_node.left
      end
    end

    unless node.right.nil?
      loop do
        break if loop_node.right.nil?
        max_height += 1
        loop_node = loop_node.right
      end
    end

    [min_height, max_height].max
  end

  def self.max(node)
    return nil if node.nil?
    return node if node.right.nil?

    BinarySearchTree.max(node.right)
  end

  def self.min(node)
    return nil if node.nil?
    return node if node.left.nil?

    BinarySearchTree.min(node.left)
  end

  def self.delete_min!(node)
    return nil unless node
    return node.right unless node.left

    node.left = BinarySearchTree.delete_min!(node.left)
    node
    # return nil if node.nil?


    # parent_node = nil
    # min_node = node
    #
    # unless node.left.nil?
    #   loop do
    #     parent_node = min_node
    #     min_node = node.left
    #     break if min_node.left.nil?
    #   end
    # end
    #
    # unless min_node.right.nil?
    #   parent_node.left = min_node.right
    # end

  end

  def self.find_parent(node, value)

    loop do
      unless node.left.nil?
        return [node, node.left] if node.left.value == value

        if value <= node.value
          node = node.left
        end
      end

      unless node.right.nil?
        return [node, node.right] if node.right.value == value

        if value > node.value
          node = node.right
        end
      end
    end

  end

  def self.delete!(node, value)
    return nil if node.nil?
    return node = nil if node.value == value
    # we need to keep track of the parent node
    parent, child = BinarySearchTree.find_parent(node, value)
    # 3 cases:
    # 1. no children
    if child.right.nil? && child.left.nil?
      return parent.left == child ?
              parent.left = nil : parent.right = nil
    end

    min_el = BinarySearchTree.min(child)
    min_el_parent = BinarySearchTree.find_parent(min_el, min_el.value)

    if parent.left == child
      # 2. one child
      if (child.right.nil? && !child.left.nil?)
        parent.left = child.left
      elsif (!child.right.nil? && child.left.nil?)
        parent.left = child.right
      # 3. two children!
      elsif !child.right.nil? && !child.left.nil?
        min_el.left, min_el.right = child.left, child.right

        parent.left = min_el

        min_el_parent.left == min_el ?
          min_el_parent.left = nil :
          min_el_parent.right = nil
      end

    elsif parent.right == child
      # 2. one child
      if (!child.right.nil? && child.left.nil?)
        parent.right = child.right
      elsif (child.right.nil? && !child.left.nil?)
        parent.right = child.left
      # 3. two children!
      elsif !child.right.nil? && !child.left.nil?
        min_el.left, min_el.right = child.left, child.right

        parent.right = min_el

        min_el_parent.left == min_el ?
          min_el_parent.left = nil :
          min_el_parent.right = nil
      end
    end

  end















end

File: whiteboard problems/bst.js


// Binary Search Tree (non-AVL/ non balancing)
// Space - O(n)
// Insert - O(log n)
// Delete - O(log n)
// pre-order
// post-order
// in-order

class BSTNode {
  constructor(val) {
    this.val = val
    this.left = null
    this.right = null
  }
}


class BST {
  constructor() {
    this.root = null
  }

  find(val) {
    if (this.root === null) return "Tree is empty";

    let node = this.root;
    let searching = true;

    while (searching) {
      if (val < node.val) {
        node = node.left;
      } else if (val > node.val) {
        node = node.right;
      }

      if (node.val === val) {
        return node;
      } else if (node.left === null && node.right === null) {
        return "node not found";
      }
    }

    return node;

  }

  insert(val) {
    if (this.root === null) return this.root = new BSTNode(val);

    let node = this.root;
    let searching = true;

    while (searching) {
      if (val < node.val) {
        node.left === null ? node.left = new BSTNode(val) : node = node.left;
      } else if (val > node.val) {
        node.right === null ? node.right = new BSTNode(val) : node = node.right;
      }

      if (node.val === val) {
        searching = false;
      }

    }
    return node;
  }


  inOrder(node = this.root) {
    let rest;
    let order = [];

    if (node.left === null) {
      order.push(node.val);
      return order;
    } else {
      rest = this.inOrder(node.left);
      order = order.concat(rest);
    }
    order.push(node.val);

    if (node.right) {
      order = order.concat(this.inOrder(node.right));
    }

    return order;
  }


  preOrder(node = this.root) {

    let order = [];
    let rest;
    order.push(node.val);

    if (node.left) {
      rest = this.preOrder(node.left);
      order = order.concat(rest);
    }

    if (node.right) {
      rest = this.preOrder(node.right);
      order = order.concat(rest);
    }
    return order;
  }


  postOrder(node = this.root) {

    let order = [];
    let rest;

    if (node.left) {
      rest = this.postOrder(node.left);
      order = order.concat(rest);
    }

    if (node.right) {
      rest = this.postOrder(node.right);
      order = order.concat(rest);
    }
    order.push(node.val);

    return order;
  }




  delMin(node = this.root) {
    if (!node.left) return node;

    let searching = true;
    let parent = node;
    let min;

    while (searching) {
      if (!node.left.left) {
        searching = false;
        min = node.left;
        node.left = null;
        return min;
      } else {
        node = node.left;
      }
    }

  }

  delete(val) {
    let min;
    if (this.root.val === val) {
      min = this.delMin(this.root.right);
      [min.left, min.right] = [this.root.left, this.root.right];
      this.root = min;
      return;
    }

    let searching = true;
    let node = this.root;
    let childIdx = 1;
    let parent;
    let children = [node.left, node.right]

    while (searching) {
      if (val < node.val) {
        parent = node;
        node = node.left;
        childIdx = 0;
      } else if (val > node.val) {
        parent = node;
        node = node.right;
        childIdx = 1;
      } else if (val === node.val) {
        searching = false;
      }
    }

    let singleChild;
    if (node.left && node.right) {
      // if node to delete has two children, find the min val on right
      min = this.delMin(node.right);
      if (node.right === min) {
        min.left = node.left;
      } else {
        min.left = node.left;
        min.right = node.right;
      }
      childIdx === 0 ? parent.left = min : parent.right = min

    } else if (node.left || node.right) {
      // if there is only one child on the node to be deleted
      singleChild = node.left ? node.left : node.right
      childIdx === 0 ? parent.left = singleChild : parent.right = singleChild

    } else {
      // if it has no children nodes at all
      childIdx === 0 ? parent.left = null : parent.right = null
    }

  }

}

let bst = new BST;
bst.insert(6);
bst.insert(4);
bst.insert(3);
bst.insert(5);
bst.insert(10);
bst.insert(9);
bst.insert(11);


const isBst = (node) => {
  if (!node.left && !node.right) return true;

  if (node.left) {
    if (node.left.val > node.val) {
      return false;
    } else {
      dive = isBst(node.left);
    }
  }

  if (node.right) {
    if (node.right.val < node.val) {
      return false;
    } else {
      dive = isBst(node.right);
    }
  }

  return dive;

}
isBst(bst.root);

// Find the lowest common ancestor of two nodes in a binary search tree.
// Write the function in JS. Assume I give you both the root and the
// two nodes.






//
